File: DSA_practice/All_Sorting.c
#include<stdio.h>

//Bubble sort
void Bubble(int A[], int n){
    int flag, temp;
    for(int i=0; i<n-1; i++){
        flag =0;
        for(int j=0; j<n-1-i; j++){
            if(A[j]>A[j+1]){
                temp = A[j];
                A[j] = A[j+1];
                A[j+1] = temp;
                flag=1;
            }
        }
        if(flag==0) return;
    }
}

//Insertion sort
void Insertion(int A[], int n){
    int i, j, x;
    for(i=1; i<n; i++){
        j=i-1;
        x = A[i];
        while(j>-1 && A[j]>x){
            A[j+1] = A[j];
            j--;
        }
        A[j+1] = x;
    }
}

//Selection sort
void Selection(int A[], int n){
    int i, j, k;
    for(i=0; i<n-1; i++){
        for(j=k=i; j<n; j++){
            if(A[j]<A[k]){
                k=j;
            }
        }
        int temp = A[i];
        A[i] = A[k];
        A[k] = temp;
    }
}

void Swap(int *a, int *b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

//Partition function for Quick Sort
int Partition(int A[], int l, int h){
    int pivot = A[l];
    int i=l, j=h;
    do{
        do{i++;}while(A[i]<=pivot);
        do{j--;}while(A[j]>pivot);
        if(i<j) Swap(&A[i], &A[j]);
    }while(i<j);
    int temp = A[l];
    A[l] = A[j];
    A[j] = temp;
    return j;
}

//Quick Sort
void Quick(int A[], int l, int h){
    int j;
    if(l<h){
        j = Partition(A, l, h);
        Quick(A, l, j);
        Quick(A, j+1, h);
    }
}


void merge(int arr[], int low, int mid, int high){
    int i=low, j=mid, k=0;
    int aux[high];
    while(i<mid && j<high){
        if(arr[i]<arr[j]){
            aux[k++] = arr[i++];
        }else aux[k++] = arr[j++];
    }

    while(i<mid){
        aux[k++] = arr[i++];
    }

    while(j<high){
        aux[k++] = arr[j++];
    }

    for(i=0; i<high; i++){
        arr[i] = aux[i];
    }
}

void mergeSort(int arr[], int low, int high){
    if(low<high){
        int mid = (low+high)/2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid+1, high);
        merge(arr, low, mid, high);
    }
}

int main(){
    int Arr[11] = {10,3,2,5,9,0,8,4,1,7,6};
    for(int i=0; i<11; i++){
        printf("%d ", Arr[i]);
    }
    printf("\n");
    mergeSort(Arr, 0, 11);
    for(int i=0; i<11; i++){
        printf("%d ", Arr[i]);
    }
    // Insertion(Arr, 11);
    // for(int i=0; i<11; i++){
    //     printf("%d\t", Arr[i]);
    // }
    return 0;
}

File: CompetitiveProgramming/All_Sorting.java
public class All_Sorting {
    public static void main(String[] args) {
        int arr[] = { 3, 1, 5, 9, 10, 0, -2 };
        // bubbleSort(arr);
        // insertionSort(arr);
        // selectionSort(arr);
        printArray(arr);
        mergeSort(arr, 0, 6);
        printArray(arr);
    }

    public static void bubbleSort(int arr[]) {
        System.out.println("Bubble Sort: ");
        printArray(arr);
        for (int i = 0; i < arr.length; i++) {
            boolean flag = false;
            for (int j = 0; j < arr.length; j++) {
                if (arr[j] > arr[i]) {
                    flag = true;
                    arr[i] = arr[i] ^ arr[j];
                    arr[j] = arr[i] ^ arr[j];
                    arr[i] = arr[i] ^ arr[j];
                }
            }
            if (!flag)
                break;
        }
        printArray(arr);
    }

    public static void insertionSort(int arr[]) {
        System.out.println("Insertion Sort: ");
        printArray(arr);
        for (int i = 1; i < arr.length; i++) {
            int temp = arr[i];
            int ptr = i - 1;
            while (ptr >= 0 && temp < arr[ptr]) {
                arr[ptr + 1] = arr[ptr];
                ptr = ptr - 1;
            }
            arr[ptr + 1] = temp;
        }
        printArray(arr);
    }

    public static void selectionSort(int arr[]) {
        System.out.println("Selection Sort: ");
        printArray(arr);
        for (int i = 0; i < arr.length; i++) {
            int min = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[min] > arr[j])
                    min = j;
            }
            if (min != i) {
                int temp = arr[min];
                arr[min] = arr[i];
                arr[i] = temp;
            }
        }
        printArray(arr);
    }

    public static void merge(int arr[], int low, int mid, int high){
        int aux[] = new int[high+1];
        int i=low, j=mid+1, k=0;
        while(i<=mid && j<=high){
            if(arr[i]<=arr[j]){
                aux[k++] = arr[i++];
            }
            else{
                aux[k++] = arr[j++];
            }
        }

        while(i<=mid){
            aux[k++] = arr[i++];
        }

        while(j<=high){
            aux[k++] = arr[j++];
        }
        k=0;
        for(i=low; i<=high; i++){
            arr[i] = aux[k++];
        }
    }

    public static void mergeSort(int arr[], int low, int high) {
        if (low < high) {
            int mid = (low+high)/2;
            mergeSort(arr, low, mid);
            mergeSort(arr, mid+1, high);
            merge(arr, low, mid, high);
        }
    }

    public static void printArray(int arr[]) {
        System.out.println();
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
